REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [1]:
import pandas as pd

In [2]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [3]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [4]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [5]:
print(len(bus))

6315


#### Question 2b: How many unique business IDs are there?  

In [6]:
print(len(pd.unique(bus['business_id'])))

6315


#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [25]:
print('The five most common businessses are: ',bus['name'].value_counts().head())

print('There are ',bus['city'].value_counts()['San Francisco'] , 'businesses in San Francisco')



The five most common businessses are:  name
STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: count, dtype: int64
There are  6315 businesses in San Francisco


## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [15]:
bus.dtypes

business_id       int64
name             object
address          object
city             object
state            object
postal_code      object
latitude        float64
longitude       float64
phone_number     object
dtype: object

#### Question 3b

What are the unique values of postal_code?

In [16]:
bus.postal_code.unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [17]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [71]:
vb = bus[bus['postal_code'].isin(validZip)]
print(len(vb))

5999


## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [31]:
print(vb[pd.isnull(vb['longitude'])])
print(len(vb[pd.isnull(vb['longitude'])]),'businesses are missing longitude values')

      business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Fr

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [101]:
# create dataframes
zip_table = pd.DataFrame(validZip, columns=['postal_code'])

valid_long = pd.DataFrame(bus.groupby(['postal_code']).count(),columns = ['longitude'])
valid_long.rename(columns={'longitude':'not_null_long'}, inplace=True)

not_valid_long = pd.DataFrame(vb.groupby(['postal_code']).size(),columns = ['longitude'])
not_valid_long.rename(columns={'longitude':'total_long'}, inplace=True)


#join dataframes
join1 = pd.merge(zip_table, valid_long, on='postal_code',how='left')
join2 = pd.merge(join1, not_valid_long, on='postal_code',how='left')
join2['null_long']=(join2['total_long']-join2['not_null_long'])


print(join2.drop(columns={'total_long'}))

   postal_code  not_null_long  null_long
0        94102            262        196
1        94103            306        252
2        94104             74         59
3        94105            126        100
4        94107            204        247
5        94108            142         82
6        94109            237        146
7        94110            334        251
8        94111            167        115
9        94112            131         59
10       94114            122        108
11       94115            162         76
12       94116             58         35
13       94117            130         67
14       94118            143        100
15       94121            106         55
16       94122            147        123
17       94123            116         53
18       94124             76         88
19       94127             43         28
20       94131             34         14
21       94132             69         65
22       94133            282        135
23       94134  

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [103]:
join2['prop_missing']=join2['null_long']/(join2['total_long'])

print(join2.sort_values(by='prop_missing',ascending=False).head())                                                
                                          

   postal_code  not_null_long  total_long  null_long  prop_missing
4        94107            204         451        247      0.547672
18       94124             76         164         88      0.536585
21       94132             69         134         65      0.485075
10       94114            122         230        108      0.469565
16       94122            147         270        123      0.455556


# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [108]:
ins.head()
print(len(ins))
print(len(pd.unique(ins['business_id'])))

15430
5730


#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [122]:
ins['type'].value_counts()

ins2 = ins[ins['type'].map(ins['type'].value_counts())> 10]

print(ins2)

       business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15429 rows x 4 columns]


#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [171]:
ins2['date'] = ins2['date'].astype('string')
ins2['year']=(ins2['date'].str.slice(0,4))
print(ins2)
#ins2.drop(['year_str','date_str'], axis=1)


       business_id  score      date     type  year    year_str  date_str
0               10     82  20160503  routine  2016  20160503.0  20160503
1               10     94  20140729  routine  2014  20140729.0  20140729
2               10     92  20140114  routine  2014  20140114.0  20140114
3               19     94  20160513  routine  2016         nan  20160513
4               19     94  20141110  routine  2014         nan  20141110
...            ...    ...       ...      ...   ...         ...       ...
15425        89515     98  20161103  routine  2016         nan  20161103
15426        89701    100  20161206  routine  2016         nan  20161206
15427        89790     96  20161129  routine  2016         nan  20161129
15428        89803    100  20161215  routine  2016         nan  20161215
15429        89900     93  20161206  routine  2016         nan  20161206

[15429 rows x 7 columns]


C:\Users\lssch\AppData\Local\Temp\ipykernel_10940\2782048107.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ins2['date'] = ins2['date'].astype('string')
C:\Users\lssch\AppData\Local\Temp\ipykernel_10940\2782048107.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ins2['year']=(ins2['date'].str.slice(0,4))


#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [172]:
ins2['year'].value_counts()
ins3 = ins2[ins2['year'].map(ins2['year'].value_counts())>= 50]
print(ins3)

       business_id  score      date     type  year    year_str  date_str
0               10     82  20160503  routine  2016  20160503.0  20160503
1               10     94  20140729  routine  2014  20140729.0  20140729
2               10     92  20140114  routine  2014  20140114.0  20140114
3               19     94  20160513  routine  2016         nan  20160513
4               19     94  20141110  routine  2014         nan  20141110
...            ...    ...       ...      ...   ...         ...       ...
15425        89515     98  20161103  routine  2016         nan  20161103
15426        89701    100  20161206  routine  2016         nan  20161206
15427        89790     96  20161129  routine  2016         nan  20161129
15428        89803    100  20161215  routine  2016         nan  20161215
15429        89900     93  20161206  routine  2016         nan  20161206

[15391 rows x 7 columns]


Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [173]:
year_2016 = ins3[ins3['year'].isin(['2016'])]
print(year_2016)

       business_id  score      date     type  year    year_str  date_str
0               10     82  20160503  routine  2016  20160503.0  20160503
3               19     94  20160513  routine  2016         nan  20160513
6               24     98  20161005  routine  2016         nan  20161005
7               24     96  20160311  routine  2016         nan  20160311
10              31     84  20160816  routine  2016         nan  20160816
...            ...    ...       ...      ...   ...         ...       ...
15425        89515     98  20161103  routine  2016         nan  20161103
15426        89701    100  20161206  routine  2016         nan  20161206
15427        89790     96  20161129  routine  2016         nan  20161129
15428        89803    100  20161215  routine  2016         nan  20161215
15429        89900     93  20161206  routine  2016         nan  20161206

[5839 rows x 7 columns]


### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [174]:
merged_table = pd.merge(bus, year_2016, on='business_id',how='left')
print(merged_table.head())

   business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4           31      NORMAN'S ICE CREAM AND FREEZES   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
4           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   

    longitude  phone_number  score      date     type  year    year_str  \
0 -122.403816  +14154217044   82.0  20160503  routine  2016  20160503.0   
1 -122.421547  +14157763262   94.0  20160513

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [175]:
print(merged_table[['name','address','score']].sort_values(by='score').head(20)) 

                             name              address  score
4892      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    52.0
7324                    POKI TIME     2101 LOMBARD ST    55.0
6281                 CRAZY PEPPER   2257 SAN JOSE AVE    55.0
4873                   UNCLE CAFE       65 WAVERLY PL    55.0
3008             HING WANG BAKERY        339 JUDAH ST    56.0
5627                   GOLDEN WOK   295 B ORIZABA AVE    56.0
5383                   TOP'S CAFE  2460 SAN BRUNO AVE    57.0
6067    L & G VIETNAMESE SANDWICH         602 EDDY ST    57.0
836                  RED A BAKERY      634 CLEMENT ST    57.0
4209      SAN TUNG RESTAURANT LLC      1031 IRVING ST    57.0
549           MA'S DIM SUM & CAFE      1315 POWELL ST    57.0
4903  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    57.0
6645    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    58.0
624           JOANIE'S DINER INC.   1329 COLUMBUS AVE    59.0
6938         PAK NAM THAI CUISINE       655 LARKIN ST    59.0
3563    

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!